akvo/rsr/usecases/run_iati_validation_jobs.py

```diff
@@ -0,0 +1,238 @@
+        'content': json.dumps(check_result.data, indent=2),
+        'mimetype': 'application/json'
+    }]
+    rsr_send_mail(
+        email_recipients,
+        subject='iati_validation/activity_error_subject.txt',
+        message='iati_validation/activity_error_message.txt',
+        msg_context={'project': project, 'api_result': validator_result, 'rsr_result': check_result},
+        attachments=attachments
+    )
+
+
+def get_iati_activity_xml_doc(project: Project) -> bytes:
+    return etree.tostring(etree.ElementTree(IatiXML([project]).iati_activities))
+
+
+def run_iati_organisation_validation_job(scheduled_at: datetime = now()):
+    if not rate_limiter.is_allowed():
+        return
+    pending_jobs = get_pending_organisation_jobs(scheduled_at)
+    if not pending_jobs.exists():
+        return
+    job = pending_jobs.first()
+    organisation = job.organisation
+    job.mark_started()
+    try:
+        validator_result = validator.validate(get_iati_organisation_xml_doc(organisation))
+        job.mark_finished()
+        process_organisation_validation_results(organisation, validator_result)
+    except IATIValidatorTimeoutException:
+        pass  # retry on next round
+    except IATIValidatorException:
+        logger.exception(f'Failed to run IATI validator for organisation: {organisation.name}')
+
+
+def get_pending_organisation_jobs(scheduled_at: datetime = now()):
+    # Add 10 seconds to avoid race conditions for the failed jobs
+    timeout_at = now() - timedelta(seconds=VALIDATOR_TIMEOUT + 10)
+    return IatiOrganisationValidationJob.objects\
+        .select_related('organisation')\
+        .filter(scheduled_at__lte=scheduled_at, finished_at__isnull=True)\
+        .exclude(Q(attempts__gte=VALIDATOR_MAX_ATTEMPTS) | Q(started_at__gte=timeout_at))
+
+
+def process_organisation_validation_results(organisation: Organisation, validator_result: IATIValidationResult):
+    if validator_result.error_count == 0 and validator_result.warning_count == 0:
+        return
+    message = f'IATI validation error for organisation: {organisation.name}\n'\
+        f'errors: {validator_result.error_count}\n'\
+        f'warnings: {validator_result.warning_count}'
+    logger.info(message)
+    email_recipients = getattr(settings, 'IATI_ORGANISATION_VALIDATION_ERROR_RECIPIENTS', [])
+    if not email_recipients:
+        return
+    attachments = [{
+        'filename': 'iati-validator-api-result.json',
+        'content': json.dumps(validator_result.data, indent=2),
+        'mimetype': 'application/json'
+    }]
+    rsr_send_mail(
+        email_recipients,
+        subject='iati_validation/organisation_error_subject.txt',
+        message='iati_validation/organisation_error_message.txt',
+        msg_context={'organisation': organisation, 'result': validator_result},
+        attachments=attachments
+    )
+
+
+def get_iati_organisation_xml_doc(organisation: Organisation, context: Dict[str, str] = {}) -> bytes:
```

Comment:
The default argument should not be mutable, otherwise future calls of the function might get the modified default

Reply:
Fixed